Addition of lenses for PParams

lehins opened the pull request:
# Description

Addition of  `EraPParams`, `AlonzoEraPParams` and `BabageEraPParams` type classes with lenses that can access and manipulate both `PParams` and `ParamsUpdate`. Each type class requires implementation lenses for the `PParamsHKD` type, which are later used for defining specific lenses for PParams and PParamsUpdate respectfully

HKD types `ShelleyPParams`, `AlonzoPParams` and `BabbagePParams` now become internal, for the most part and the `PParams`+`PParamsUpdate` are the newtype wrappers around the era specific internal HKD type. All operations on PParams should be done with lenses from now on.

Other notable things in this PR:

* Fixes `Ord` instances for `AlonzoPParams` and `BabbagePParams` by introducing an internal `OrdExUniits type.
* Moves around `To/FromJSON` instances to prevent the from being orphan
* Moved `CostModels` into `Cardano.Ledger.Alozno.Scripts` from `Cardano.Ledger.Alozno.PParams`
* Switch all functions that accepted PParams as an argument, but only used protocol version, to accept only protocol version. Most notable examples are functions that do conditional checking on the hardfork
* Makes `PoolEnv` fields strict
* Adds a data type `ByronEra` that has an `Era` instance. This is done mostly for completeness, but also to allow getting protocol version from the type level. Also this allowed `PreviousEra` type family to be moved into `Era` type class.
* Moved all class definitions {EraName}Era{Type}, into the `Cardano.Ledger.{EraName}.Core` module. It was discussed before, but wasn't fully implemented.

Fixes #2927 

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (which can be run with `scripts/fourmolize.sh`
- [x] Self-reviewed the diff

Soupstraw committed:
Redesign how PParams are defined in all eras. Remove `HasField` usage.

* Added `EraPParams` `AlonzoEraPParams` and `BabageEraPParams` with lenses
  that can access and manipulate both `PParams` and `ParamsUpdate` for each era.
* Introduced `upgradePParams` and `downgradePParams`, which loosely
  correspond to `extendPP` and `retractPP`.
* Removed all occurrences of HasField from all the eras.
* Extracted `Era` related code and `TranslateEra` related code from
  `Cardano.Ledger.Core` into `Cardano.Ledger.Core.Era` and
  `Cardano.Ledger.Core.Translation` respectfully
* Moved `PreviousEra` type family into `Era` type class.
* Defined an `Era` instance for new `ByronEra` type
* Concrete PParams types for each era had their fields renamed to follow
  the convention used throughout the codebase
* Adjust transaction generation in the travce generator to work for
  eras beyond Mary
* Got rid of `ShelleyTest` and `UsesPP` type synonyms

lehins committed:
Workaround a ghc bug that was fixed in ghc-9.2

It is acceptable to turn off the warning for the whole package
for ghc-8.10 since we are developing with ghc-9.2 and we have CI
for both of those GHC versions, so we will definitely catch any
name shadowing warnings that aren't related to the bug.

lehins committed:
Fix field naming for `FreeVars` and ensure correct `Coin` type is used

lehins committed:
Rename `RewardSnapShot` field `rewprotocolVersion` -> `rewProtocolVersion`

Soupstraw commented on CHANGELOG.md:
```diff
@@ -12,6 +12,9 @@ in the naming of release branches.
 
 ### Added
 
+- Added type classes: `EraPParams`, `AlonzoEraPParams` and `BabageEraPParams` with lenses
```
```suggestion
- Added type classes: `EraPParams`, `AlonzoEraPParams` and `BabbageEraPParams` with lenses
```

lehins replied:
Done

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs:
```diff
@@ -112,165 +144,46 @@ instance FromCBOR AlonzoGenesis where
+        , agMaxBlockExUnits
+        , agMaxValSize
+        , agCollateralPercentage
+        , agMaxCollateralInputs
```
This could be done more compactly with sequential application:
```haskell
instance FromJSON AlonzoGenesis where
  parseJSON = Aeson.withObject "Alonzo Genesis" $ \o ->
    AlonzoGenesis
      <$> o .: "lovelacePerUTxOWord"
      <*> o .: "costModels"
      <*> ...
```

lehins replied:
I don't like this style cause it is easy to mess up the order. For example if the order of fields has changed in the type itself, or I just happen to mess up the order in the implementation then compiler is not gonna help me in any way:

```diff
instance FromJSON AlonzoGenesis where
  parseJSON = Aeson.withObject "Alonzo Genesis" $ \o ->
    AlonzoGenesis
      <$> o .: "lovelacePerUTxOWord"
      <*> o .: "costModels"
      <*> ...
-      <*> o .: "agMaxTxExUnits"
-      <*> o .: "agMaxBlockExUnits"
+      <*> o .: "agMaxBlockExUnits"
+      <*> o .: "agMaxTxExUnits"
      <*> ...
```


That being said, I could use the record wildcards at the end, which would make just as compact as the version above, but still safe:
```haskell
    return AlonzoGenesis {..}
```

I'll do that when there will be any more suggestions from others

lehins replied:
Done. Thank you.

lehins committed:
Fix a wrong assumption about how caching works

* According to Github documentation cache action will already
automatically look into `master` branch cache, however it will never
look into caches for other branches. So, previous logic was a bit faulty
* Enable trigger on pushes to master, this way we can have a build after
each PR is merged.

Soupstraw commented on eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs:
```diff
@@ -669,7 +654,7 @@ txOutDataHash = strictMaybeToMaybe . getDataHashBabbageTxOut
 txOutScript :: BabbageTxOut era -> Maybe (Script era)
 txOutScript = strictMaybeToMaybe . getScriptBabbageTxOut
-{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBabbageTxOut`" #-}
+{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBa\bbageTxOut`" #-}
```
```suggestion
{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBabbageTxOut`" #-}
```

lehins replied:
Thank you. Fixed

Soupstraw commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -272,21 +267,19 @@ addTokens ::
-   in if Val.coin v < scaledMinDeposit v (getField @"_minUTxOValue" pparams)
-        then addTokens proxy (txOut :<| tooLittleLovelace) pparams ts os
-        else Just $ tooLittleLovelace >< addValToTxOut @era (MaryValue 0 ts) txOut <| os
+  if txOut ^. coinTxOutL < getMinCoinTxOut pparams txOut
```
Nice, much more readable

Soupstraw commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs:
```diff
@@ -21,7 +21,7 @@ import Cardano.Ledger.Core
 import Cardano.Ledger.Shelley.Delegation.Certificates (DCert)
 import Cardano.Ledger.Shelley.Era (ShelleyEra)
-import Cardano.Ledger.Shelley.PParams (Update)
+import Cardano.Ledger.Shelley.PParams
```
Unnecessary change?

lehins replied:
Yep. Fixed.

Soupstraw commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs:
```diff
@@ -21,9 +21,9 @@ import qualified Cardano.Crypto.Hashing as Hashing
+import Cardano.Ledger.Core
 import Cardano.Ledger.Crypto (Crypto (ADDRHASH))
-import Cardano.Ledger.EpochBoundary (emptySnapShots)
+import Cardano.Ledger.EpochBoundary
```
Unnecessary change

lehins replied:
Yep, done.

Soupstraw commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs:
```diff
@@ -241,12 +245,12 @@ blockEx2 =
 expectedStEx2 ::
   forall c.
-  (ExMock c, C.UsesPP (ShelleyEra c)) =>
+  (ExMock c) =>
```
```suggestion
  ExMock c =>
```

lehins replied:
Thank you! Done

Soupstraw commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs:
```diff
@@ -259,7 +263,7 @@ expectedStEx2 =
 -- Submit an empty block to trigger adopting the genesis delegation.
 genesisDelegation2 ::
-  (ExMock c, C.UsesPP (ShelleyEra c)) =>
+  (ExMock c) =>
```
```suggestion
  ExMock c =>
```

lehins replied:
Done

Soupstraw commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Mir.hs:
```diff
@@ -149,15 +148,15 @@ txbodyEx1 pot =
   where
     aliceInitCoin = Val.inject $ Coin $ 10 * 1000 * 1000 * 1000 * 1000 * 1000
-    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> _keyDeposit ppEx)
+    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> (Coin 7))
```
```suggestion
    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> Coin 7)
```

lehins replied:
Good catch. Done

Soupstraw commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/MirTransfer.hs:
```diff
@@ -26,14 +26,15 @@ import Cardano.Ledger.Shelley.API (
   ShelleyDELEG,
  )
-import Cardano.Ledger.Shelley.PParams (ShelleyPParamsHKD (..), emptyPParams)
+import Cardano.Ledger.Shelley.Core
```
Redundant import?

Soupstraw replied:
never mind, `ppProtocolVersionL` moved to Core.

Soupstraw commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs:
```diff
@@ -59,16 +61,11 @@ import Test.Cardano.Ledger.Shelley.MultiSigExamples (
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.HUnit (Assertion, assertBool, testCase, (@?=))
-import Test.Tasty.QuickCheck (testProperty)
+import Test.Tasty.QuickCheck
```
```suggestion
import Test.Tasty.QuickCheck (testProperty)
```

Soupstraw replied:
Not sure why the imports have changed in this file..

lehins replied:
That's because I also ended up needing `Property`,  `discard`, `(===)`, but I wasn't sure what else I might need, thus ended up removing it temporarily, but forgot to make it explicit again.

Fixed

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Language.hs:
```diff
@@ -96,7 +129,8 @@ instance IsLanguage 'PlutusV2 where
   | fromSLanguage thisLanguage == lang = pure thisLanguage
-  | otherwise = fail $ "Plutus language mismatch. Expected " ++ show thisLanguage ++ ", but got: " ++ show lang
+  | otherwise =
+      fail $ "Plutus language mismatch. Expected " ++ show thisLanguage ++ ", but got: " ++ show lang
```
Unnecessary change

lehins replied:
This one actually was need, since a line was way too long. Taking this part to another line shaved off 10 characters :slightly_smiling_face: 
Its is still a bit too long, I'll switch it to a multi-line.

Done

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -561,39 +649,65 @@ encodeLangViews views = encodeMapLen n <> foldMap encPair ascending
-instance ToExpr (AlonzoPParamsHKD StrictMaybe era)
-
-instance ToExpr (AlonzoPParamsHKD Identity era)
+-- | Given the missing pieces Turn a ShelleyPParams into a AlonzoPParams
```
```suggestion
-- | Given the missing pieces, turn a ShelleyPParams into a AlonzoPParams
```

lehins replied:
Done

JaredCorduan commented on eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Examples/Consensus.hs:
```diff
@@ -52,13 +50,12 @@ import qualified Data.Map.Strict as Map
 import Data.Proxy (Proxy (..))
 import qualified Data.Sequence.Strict as StrictSeq
 import qualified Data.Set as Set
+import Lens.Micro
```
is this a library that we also want to allow open imports for? :thinking:

lehins replied:
Yeah, for sure, since now it became omnipresent.

JaredCorduan commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -3,34 +3,83 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE GADTs #-}
+{-# LANGUAGE InstanceSigs #-}
```
:heart:

lehins replied:
Yeah. Neat trick. It is need to get to the `f` type parameter in the instance definition:
```haskell
  hkdMaxTxExUnitsL :: forall f. HKDFunctor f => Lens' (PParamsHKD f (ConwayEra c)) (HKD f ExUnits)
  hkdMaxTxExUnitsL =
    lens (hkdMap (Proxy @f) unOrdExUnits . bppMaxTxExUnits) $ \pp x ->
      pp {bppMaxTxExUnits = hkdMap (Proxy @f) OrdExUnits x}
```

JaredCorduan commented on eras/mary/impl/src/Cardano/Ledger/Mary/Translation.hs:
```diff
@@ -117,7 +114,7 @@ instance Crypto c => TranslateEra (MaryEra c) ShelleyTxOut where
     pure $ TxOutCompact (coerce addr) (translateCompactValue cfval)
 
-instance Crypto c => TranslateEra (MaryEra c) UTxO where
+instance (Crypto c) => TranslateEra (MaryEra c) UTxO where
```
I thought you were the destroyer of extraneous parenthesis :laughing:

lehins replied:
I am! I have no idea how I could commit such an atrocity!
:smile: 

Fixed

JaredCorduan commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs:
```diff
@@ -449,26 +442,6 @@ validateVerifiedWits tx =
-   in failureUnless (nullWitHashes missingWitnesses) $
-        MissingVKeyWitnessesUTXOW missingWitnesses
--}
-
```
:pray:

JaredCorduan reviewed (approved):
This all looks amazing, thank you!

You don't have to do this, just an idea: maybe we could add a section to the `CONTRIBUTING.md` doc about open imports, forbidden them except for `Core` modules, and maybe lenses?
